### PR TITLE
add Detail

### PR DESCRIPTION
**Pull request by thinhnguyen91**
add Detail, Temp, change LocationVC, CountryVC, CityVC, RegisterVC, DetailViewController

### CODE REVIEW

File: Weather/Controllers/Define/Define.swift [hunk 1]
```diff
@@ -19,3 +19,9 @@ struct LinkCITY {
     static let urlCity = "https://api.meetup.com/cities?&sign=true&photo-host=public&country="
     static let urlCityKey = "&page=500&key=27527d4633352c7b1624101f5e6467"
 }
+
+struct LinkWEATHER {
+    static let urlWather = "http://api.openweathermap.org/data/2.5/forecast/daily?q="
```
Comment (at-phuongvnc) [inline Weather/Controllers/Define/Define.swift]: wrong name

File: Weather/Controllers/HomeViewController/HomeViewController.swift [hunk 2]
```diff
@@ -56,7 +56,7 @@ class HomeViewController: BaseViewController {
         alertController.addAction(noAction)
 
         let yesAction = UIAlertAction(title: "Yes", style: .Default) { (action) in
-            let myRegisteeVC = RegisterViewController()
+            let myRegisteeVC = RegisterViewController(country: "", city: "")
```
Comment (at-phuongvnc) [inline Weather/Controllers/HomeViewController/HomeViewController.swift]: you can use init(country:String = "",city:String = "")

File: Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift [hunk 3]
```diff
@@ -10,7 +10,7 @@ import UIKit
 import ObjectMapper
 
 protocol CountryViewControllerDelegate {
-    func showData(name: String, cell: LocationCell)
+    func showData(nameCountry: String)
     func show(isCheck: Bool)
```
Comment (at-phuongvnc) [inline Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift]: show ?

File: Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift [hunk 4]
```diff
@@ -22,13 +22,11 @@ private struct Options {
 class CountryViewController: BaseViewController {
     private var weather = WebService()
     var delegate: CountryViewControllerDelegate?
-    var cell: LocationCell?
     var isCheck = false
     private var altspelling = ""
```
Comment (at-phuongvnc) [inline Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift]: use `camel case`

File: Weather/Controllers/LocationViewController/LocationViewController.swift [hunk 5]
```diff
@@ -9,30 +9,41 @@
...
+        var name: String
+        init(title: String, name: String) {
+            self.title = title
+            self.name = name
+        }
+    }
+    private var array = [CountryDetail]()
     private var myLocationCell = LocationCell()
     var isCheckCountry = false
     private var altspell = ""
```
Comment (at-phuongvnc) [inline Weather/Controllers/LocationViewController/LocationViewController.swift]: use `camel case`

File: Weather/Controllers/LocationViewController/LocationViewController.swift [hunk 6]
```diff
@@ -65,33 +75,34 @@ extension LocationViewController: UITableViewDelegate, UITableViewDataSource {
...
-        let cell = self.tableView.dequeueReusableCellWithIdentifier("locationCell") as? LocationCell
-        let item = array[indexPath.row]
-        cell?.nameLabel.text = item
-        if indexPath.row == 1 {
-            cell?.hidden = true
-        }
-        if isCheckCountry {
-            if indexPath.row == 1 {
-                cell?.hidden = false
+        let cell = tableView.dequeueReusableCellWithIdentifier("locationCell")! as UITableViewCell
```
Comment (at-phuongvnc) [inline Weather/Controllers/LocationViewController/LocationViewController.swift]: why do you cast cell to Location Cell ?

File: Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift [hunk 7]
```diff
@@ -7,27 +7,113 @@
...
+import NoOptionalInterpolation
 
 class DetailViewController: BaseViewController {
+    private var weather = WebService()
+    private var details = [Detail]()
+    private var detail: Detail?
+    var country: String!
+    var city: String!
+
+    @IBOutlet weak var countryLable: UILabel!
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: missing `private`

File: Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift [hunk 8]
```diff
@@ -7,27 +7,113 @@
...
 
     }
+    init(country: String, city: String) {
+        super.init()
+        self.country = country
+        self.city = city
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: remove it

File: Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift [hunk 9]
```diff
@@ -7,27 +7,113 @@
...
+        let nib = UINib(nibName: "DetailCell", bundle: nil)
+        collectionView.registerNib(nib, forCellWithReuseIdentifier: "cell")
+    }
 
-    @IBAction private func settingAction(sender: AnyObject) {
-        let mySettingVC = SettingViewController()
-        navigationController?.pushViewController(mySettingVC, animated: true)
+    private func loadDetail() {
+        countryLable.text = city + "-" + country
+        print(city)
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: remove `print`

File: Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift [hunk 10]
```diff
@@ -7,27 +7,113 @@
...
+    func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
+
+        let cell = collectionView.dequeueReusableCellWithReuseIdentifier("cell", forIndexPath: indexPath) as? DetailCell
+
+        return cell!
+    }
+
+    func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
+
+        print(indexPath.row)
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: remove `print`

File: Weather/WebService/WebService.swift [hunk 11]
```diff
@@ -35,4 +36,14 @@ class WebService: NSObject {
                 }
         }
     }
+
+    func getURLWEATHER(city: String, completion: APIControllerCallBackWEATHER) {
+      print( LinkWEATHER.urlWather + "\(city)" + LinkWEATHER.urlDetail)
```
Comment (at-phuongvnc) [inline Weather/WebService/WebService.swift]: remove print

File: Weather/Controllers/LocationViewController/LocationViewController.swift (hunk 5 above)
Comment (thinhnguyen91) [inline Weather/Controllers/LocationViewController/LocationViewController.swift]: fix finish

File: Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift (hunk 4 above)
Comment (thinhnguyen91) [inline Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift]: fix finish

File: Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift (hunk 3 above)
Comment (thinhnguyen91) [inline Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift]: Change show -> push

File: Weather/Controllers/LocationViewController/LocationViewController.swift (hunk 6 above)
Comment (thinhnguyen91) [inline Weather/Controllers/LocationViewController/LocationViewController.swift]: fix finish

File: Weather/Controllers/Define/Define.swift [hunk 12]
```diff
@@ -11,11 +11,23 @@ struct AlertGPS {
...
+    static let urlLAST = "&page=500&key=27527d4633352c7b1624101f5e6467"
+}
+
+struct WEATHER {
+    static let urlFIRST = "http://api.openweathermap.org/data/2.5/forecast/daily?q="
+    static let urlLAST = "&mode=json&units=metric&cnt=7&appid=e28271390d6f273ba86556379815d277"
+
+}
+
+struct DateUPDATE {
```
Comment (at-phuongvnc) [inline Weather/Controllers/Define/Define.swift]: use `camel case`

File: Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift [hunk 13]
```diff
@@ -10,9 +10,9 @@ import UIKit
 import ObjectMapper
 
 protocol CountryViewControllerDelegate {
-    func showData(name: String, cell: LocationCell)
-    func show(isCheck: Bool)
-    func showSpell(altspell: String)
+    func pushData(nameCountry: String)
+    func push(isCheck: Bool)
```
Comment (at-phuongvnc) [inline Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift]: change name of `func`

File: Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift [hunk 14]
```diff
@@ -94,27 +90,24 @@ extension CountryViewController: UITableViewDelegate, UITableViewDataSource {
...
         tableView.deselectRowAtIndexPath(indexPath, animated: true)
-        if let cellrow = tableView.cellForRowAtIndexPath(indexPath) as? CountryCell {
-            let item = weathers[indexPath.row]
-            delegate?.showData(cellrow.nameCountry.text!, cell: cell!)
-            delegate?.showSpell(item.altSpelling)
-            navigationController?.popViewControllerAnimated(true)
-        }
+        let cellrow = tableView.cellForRowAtIndexPath(indexPath) as? CountryCell
+        let item = weathers[indexPath.row]
+        delegate?.pushData(cellrow!.nameCountry.text!)
```
Comment (at-phuongvnc) [inline Weather/Controllers/LocationViewController/CountryViewController/CountryViewController.swift]: add 2 func to 1 func

File: Weather/Controllers/RegisterViewController/DetailViewController/RootViewController/RootViewController.swift [hunk 15]
```diff
@@ -7,29 +7,37 @@
 //
 
 import UIKit
+import LGSideMenuController
 
-class RootViewController: UIViewController {
-
+class RootViewController: LGSideMenuController {
+    var rightViewController = SettingViewController(name: "", country: "", city: "")
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/RootViewController/RootViewController.swift]: use `init(name:String = ""...)`

File: Weather/Controllers/RegisterViewController/DetailViewController/RootViewController/RootViewController.swift [hunk 16]
```diff
@@ -7,29 +7,37 @@
...
-    /*
-    // MARK: - Navigation
-
-    // In a storyboard-based application, you will often want to do a little preparation before navigation
-    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
-        // Get the new view controller using segue.destinationViewController.
-        // Pass the selected object to the new view controller.
+    func loadMenuView(initWithRootViewController: UIViewController, style: LGSideMenuPresentationStyle) {
+        super.rootViewController = initWithRootViewController
+        rightViewController = SettingViewController(name: "", country: "", city: "")
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/RootViewController/RootViewController.swift]: use init(name:String = ""...)

File: Weather/AppDelegate/AppDelegate.swift [hunk 17]
```diff
@@ -30,6 +35,11 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         return true
     }
 
+    func changeRootViewController(navi: UINavigationController) {
+        self.window?.rootViewController = navi
+        window?.makeKeyAndVisible()
```
Comment (at-phuongvnc) [inline Weather/AppDelegate/AppDelegate.swift]: move to func `didFinishLauching..`

File: Weather/Controllers/LocationViewController/CityViewController/CityViewController.swift [hunk 18]
```diff
@@ -20,12 +20,13 @@ class CityViewController: BaseViewController {
     var weather = WebService()
```
Comment (at-phuongvnc) [inline Weather/Controllers/LocationViewController/CityViewController/CityViewController.swift]: missing `private`

File: Weather/Controllers/LocationViewController/CityViewController/CityViewController.swift [hunk 19]
```diff
@@ -20,12 +20,13 @@ class CityViewController: BaseViewController {
...
     var cell: LocationCell?
```
Comment (at-phuongvnc) [inline Weather/Controllers/LocationViewController/CityViewController/CityViewController.swift]: Do you use it ?

File: Weather/Controllers/LocationViewController/CityViewController/CityViewController.swift [hunk 20]
```diff
@@ -35,27 +36,35 @@ class CityViewController: BaseViewController {
...
     }
-    
+
     override func setUpData() {
-    
+        getJson()
+    }
+
+    // MARK: FUNC
+    private func getJson() {
```
Comment (at-phuongvnc) [inline Weather/Controllers/LocationViewController/CityViewController/CityViewController.swift]: why do you set name `getJson` ?

File: Weather/Controllers/LocationViewController/LocationViewController.swift [hunk 21]
```diff
@@ -41,94 +54,87 @@ class LocationViewController: BaseViewController {
...
     }
-
 }
 
 extension LocationViewController: UITableViewDelegate, UITableViewDataSource {
 
     func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
-
+        if indexPath.row == 1 {
+            if country.isEmpty {
```
Comment (at-phuongvnc) [inline Weather/Controllers/LocationViewController/LocationViewController.swift]: you could use `if index.row == 1 && country.isEmpty`

File: Weather/Controllers/LocationViewController/LocationViewController.swift [hunk 22]
```diff
@@ -41,94 +54,87 @@ class LocationViewController: BaseViewController {
...
-        cell?.nameLabel.text = item
-        if indexPath.row == 1 {
-            cell?.hidden = true
-        }
-        if isCheckCountry {
-            if indexPath.row == 1 {
-                cell?.hidden = false
-            }
+        let cell = tableView.dequeueReusableCellWithIdentifier("locationCell")! as? LocationCell
+        let location = Location(rawValue: indexPath.row)!
```
Comment (at-phuongvnc) [inline Weather/Controllers/LocationViewController/LocationViewController.swift]: change name of `Location enum`

File: Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift [hunk 23]
```diff
@@ -7,27 +7,198 @@
...
+        dateLable.text = "To day: " + DefineDate.defaultTimeZoneStr
+    }
+
+    override func setUpData() {
+        getJson()
+        getJsonLocation()
+    }
+
+    // MARK: FUCN
+    private func getJson() {
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: change name of func

Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: change name of func

File: Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift [hunk 24]
```diff
@@ -7,27 +7,198 @@
...
+        self.weather.getURLWEATHER(encodedCity!) { (success, result, error) in
+            self.hideLoading()
+            if success {
+                if let list = result!["list"] as? [AnyObject] {
+                    for dictionary in list {
+                        let detail = Mapper<Detail>().map(dictionary)
+                        self.details.append(detail!)
+                    }
+                    self.detail = self.details.first
+                    self.maxLable.text = "\(Int(round(self.detail!.temp!.max)))" + " °C"
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: please define `°C`

File: Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift [hunk 25]
```diff
@@ -7,27 +7,198 @@
...
+                    self.speedLable.text = "\(Int(round((self.detail!.speed) * 10)))" + " km/h"
+                    print(self.detail?.weather.icon)
+                    self.imageWeather.image = UIImage(named: "\(self.detail?.weather.icon)")
+                    self.collectionView.reloadData()
+                }
+            }
+        }
+    }
+
+    private func getJsonLocation() {
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: change name of func

File: Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift [hunk 26]
```diff
@@ -7,27 +7,198 @@
...
+                        self.maxLable.text = "\(Int(round((self.detail!.temp!.max)/10)))" + " °C"
+                        self.minLable.text = "\(Int((round(self.detail!.temp!.min)/10)))" + " °C"
+                        self.dayLable.text = "\(Int((round(self.detail!.temp!.day)/10)))" + " °C"
+                        self.humidityLable.text = "\((self.detail!.humidity)/10)" + " %"
+                        self.speedLable.text = "\(Int(round(self.detail!.speed)))" + " km/h"
+                        print(self.detail?.weather.icon)
+                        self.imageWeather.image = UIImage(named: "\(self.detail?.weather.icon)")
+                        self.collectionView.reloadData()
+                    }
+                    print(list)
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: remove `print`

File: Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift [hunk 27]
```diff
@@ -7,27 +7,198 @@
...
+                }
+            }
+        }
+    }
+
+    private func day() -> String {
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: change name of func.I don't understand it

File: Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift [hunk 28]
```diff
@@ -7,27 +7,198 @@
...
+        formatter.dateFormat = "dd"
+        let defaultTimeZoneStr = formatter.stringFromDate(date)
+        return defaultTimeZoneStr
+    }
+
+// MARK: ACTION
+    @IBAction func settingAction(sender: AnyObject) {
+        let kMainViewController = UIApplication.sharedApplication().delegate?.window!!.rootViewController as? RootViewController
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: `UIApplication.sharedApplication().delegate?.window!!` double !

File: Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift [hunk 29]
```diff
@@ -7,27 +7,198 @@
...
+        return details.count
+    }
+
+    func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
+
+        let cell = collectionView.dequeueReusableCellWithReuseIdentifier("cell", forIndexPath: indexPath) as? DetailCell
+        let index = indexPath.row
+        var userObj = Detail()
+        userObj = details[indexPath.row]
+        if index == 0 {
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/DetailViewController.swift]: define number.It is a magic number

File: Weather/Controllers/RegisterViewController/DetailViewController/SettingViewController/SettingViewController.swift [hunk 30]
```diff
@@ -8,22 +8,97 @@
...
+
+    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return 6
+
+    }
+
+    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let cell = self.tableView.dequeueReusableCellWithIdentifier("cell") as? SettingCell
+        let index = indexPath.row
+        if index == 0 {
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/SettingViewController/SettingViewController.swift]: define number

File: Weather/Controllers/RegisterViewController/DetailViewController/SettingViewController/SettingViewController.swift [hunk 31]
```diff
@@ -8,22 +8,97 @@
...
+            cell!.nameLable.text = "Name"
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/DetailViewController/SettingViewController/SettingViewController.swift]: define text.you could use enum

File: Weather/Controllers/RegisterViewController/RegisterViewController.swift [hunk 32]
```diff
@@ -8,51 +8,106 @@
...
+
 class RegisterViewController: BaseViewController {
-    var myLocationVC = LocationViewController()
-    var mySettingVC = SettingViewController()
-    var delegate: RegisterViewControllerDelegate?
+
+    private var myLocationVC = LocationViewController()
+    private var mySettingVC = SettingViewController()
+    private var weather = WebService()
+    private var service = WebService()
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/RegisterViewController.swift]: why do you have 2 varible `Webservice`

File: Weather/Controllers/RegisterViewController/RegisterViewController.swift [hunk 33]
```diff
@@ -8,51 +8,106 @@
...
+    override func setUpData() {
+        getJson()
+        getJsonLocation()
     }
     override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
         self.view.endEditing(true)
     }
 
     // MARK: FUNC
+    private func getJson() {
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/RegisterViewController.swift]: change name of `func`

File: Weather/Controllers/RegisterViewController/RegisterViewController.swift [hunk 34]
```diff
@@ -8,51 +8,106 @@
...
 
     }
+
     override func setUpUI() {
-        nameContry.text = cell?.titleLabel.text
-        nameCity.text = cell?.titleLabel.text
+        nameContry.text = city + "-" + country
         textFieldShouldReturn(nameTextField)
 
+        print(LocationManager.instance.currentLocation?.coordinate.latitude)
```
Comment (at-phuongvnc) [inline Weather/Controllers/RegisterViewController/RegisterViewController.swift]: remove `print`

**Pull request closed without merging**